File: devplaza/controllers/tutorial.go
package controllers

import (
	"devplaza/models"
	"devplaza/utils"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
)

func CreateTutorial(c *gin.Context) {
	var req CreateTutorialRequest

	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, err.Error(), nil)
		return
	}

	var tutorial = models.Tutorial{
		Title:       req.Title,
		Description: req.Desc,
		Content:     req.Content,
		Author:      req.Author,
		CoverImg:    req.CoverImg,
		Tags:        req.Tags,
		SourceLink:  req.SourceLink,
	}

	uid, ok := c.Get("uid")
	if !ok {
		utils.ErrorResponse(c, http.StatusUnauthorized, "unauthorized", nil)
		return
	}

	userId, _ := uid.(uint)
	tutorial.PublisherId = uint(userId)

	var dapp models.Dapp
	if req.DappId != 0 {
		dapp.ID = req.DappId
		if err := dapp.GetByID(); err != nil {
			utils.ErrorResponse(c, http.StatusBadRequest, "dapp not exist", nil)
			return
		}
		tutorial.DappId = &dapp.ID
	}

	// 创建数据库记录
	if err := tutorial.Create(); err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, err.Error(), nil)
		return
	}

	utils.SuccessResponse(c, http.StatusOK, "create success", tutorial)
}

func GetTutorial(c *gin.Context) {
	idParam := c.Param("id")
	id, err := strconv.Atoi(idParam)
	if err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid ID", nil)
		return
	}

	var tutorial models.Tutorial
	tutorial.ID = uint(id)

	if err = tutorial.GetByID(); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid Tutorial", nil)
		return
	}

	utils.SuccessResponse(c, http.StatusOK, "success", tutorial)
}

func QueryTutorials(c *gin.Context) {
	keyword := c.Query("keyword")
	tag := c.Query("tag")
	order := c.DefaultQuery("order", "desc")
	dappId, _ := strconv.Atoi(c.DefaultQuery("dapp_id", "0"))
	publishStatus, _ := strconv.Atoi(c.DefaultQuery("publish_status", "0"))
	userId, _ := strconv.Atoi(c.Query("user_id"))

	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "6"))

	filter := models.TutorialFilter{
		Keyword:       keyword,
		Tag:           tag,
		DappId:        uint(dappId),
		PublisherId:   userId,
		PublishStatus: publishStatus,
		OrderDesc:     order == "desc",
		Page:          page,
		PageSize:      pageSize,
	}

	tutorials, total, err := models.QueryTutorials(filter)
	if err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, err.Error(), nil)
		return
	}

	var response = QueryTutorialsResponse{
		Tutorials: tutorials,
		Page:      page,
		PageSize:  pageSize,
		Total:     total,
	}

	utils.SuccessResponse(c, http.StatusOK, "query success", response)
}

func DeleteTutorial(c *gin.Context) {
	idParam := c.Param("id")
	id, err := strconv.Atoi(idParam)
	if err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid ID", nil)
		return
	}
	var tutorial models.Tutorial
	tutorial.ID = uint(id)

	if err = tutorial.GetByID(); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid tutorial", nil)
		return
	}

	uid, ok := c.Get("uid")
	if !ok {
		utils.ErrorResponse(c, http.StatusUnauthorized, "unauthorized", nil)
		return
	}

	userId, _ := uid.(uint)
	if tutorial.PublisherId != userId {
		utils.ErrorResponse(c, http.StatusUnauthorized, "not author", nil)
		return
	}

	if err := tutorial.Delete(); err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "Failed to delete tutorial", nil)
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "delete success", nil)
}

func UpdateTutorial(c *gin.Context) {
	idParam := c.Param("id")
	id, err := strconv.Atoi(idParam)
	if err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid ID", nil)
		return
	}

	var req UpdateTutorialRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid input data", nil)
		return
	}

	var tutorial models.Tutorial
	tutorial.ID = uint(id)

	if err = tutorial.GetByID(); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid tutorial", nil)
		return
	}

	uid, ok := c.Get("uid")
	if !ok {
		utils.ErrorResponse(c, http.StatusUnauthorized, "unauthorized", nil)
		return
	}

	userId, _ := uid.(uint)
	if tutorial.PublisherId != userId {
		utils.ErrorResponse(c, http.StatusUnauthorized, "not author", nil)
		return
	}

	var dapp models.Dapp
	if req.DappId != 0 {
		dapp.ID = req.DappId
		if err := dapp.GetByID(); err != nil {
			utils.ErrorResponse(c, http.StatusBadRequest, "dapp not exist", nil)
			return
		}
		tutorial.DappId = &dapp.ID
	}

	tutorial.Title = req.Title
	tutorial.Description = req.Desc
	tutorial.Content = req.Content
	tutorial.Author = req.Author
	tutorial.SourceLink = req.SourceLink
	tutorial.CoverImg = req.CoverImg
	tutorial.Tags = req.Tags
	tutorial.PublishStatus = 1 // 编辑更新后需要重新审核

	if err := tutorial.Update(); err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "Failed to update tutorial", nil)
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "success", tutorial)
}

func UpdateTutorialPublishStatus(c *gin.Context) {
	idParam := c.Param("id")
	id, err := strconv.Atoi(idParam)
	if err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid ID", nil)
		return
	}

	var req UpdateTutorialPublishStatusRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid input data", nil)
		return
	}

	var tutorial models.Tutorial
	tutorial.ID = uint(id)

	if err = tutorial.GetByID(); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid tutorial", nil)
		return
	}

	if tutorial.PublishStatus != 1 && tutorial.PublishStatus != 2 {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid status", nil)
		return
	}

	// TODO: 2 -> 1 ?
	now := time.Now()
	tutorial.PublishStatus = req.PublishStatus
	tutorial.PublishTime = &now

	if err := tutorial.Update(); err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "Failed to update tutorial", nil)
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "success", tutorial)
}

File: src/pages/api/tutorial.ts
import { apiRequest } from './api';

// 创建教程请求参数接口
export interface CreateTutorialParams {
  title: string;
  description: string;
  content: string;
  author: string;
  source_link: string;
  cover_img: string;
  tags: string[];
  dapp_id: number;
}

export interface UpdateTutorialParams {
  title: string;
  description: string;
  content: string;
  source_link: string;
  cover_img: string;
  author: string;
  tags: string[];
  dapp_id: number;
}

export interface GetTutorialsParams {
  keyword?: string;
  tag?: string;
  order?: 'asc' | 'desc';
  page?: number;
  page_size?: number;
  publish_status?: number;
  dapp_id?: number;
  user_id?: number;
}

// 关联 User
export interface User {
  ID: number;
  email: string;
  username: string;
  avatar: string;
  github: string;
}

// 关联 Dapp
export interface Dapp {
  ID: number;
  name: string;
  description: string;
  logo: string;
  site: string;
  x: string;
}

// 教程主模型
export interface Tutorial {
  ID: number;
  title: string;
  description: string;
  content: string;
  source_link: string;
  author: string;
  cover_img: string;
  tags: string[];
  publisher_id?: number;
  publisher?: User;
  publish_time?: string;
  publish_status: number;
  dapp_id: number;
  dapp?: Dapp;
  view_count?: number;
  CreatedAt: string;
  UpdatedAt: string;
}

// 分页返回数据结构
export interface PaginatedTutorialData {
  tutorials: Tutorial[];
  page: number;
  page_size: number;
  total: number;
}

// 统一结果结构
export interface TutorialListResult {
  success: boolean;
  message: string;
  data?: PaginatedTutorialData;
}

export interface TutorialResult {
  success: boolean;
  message: string;
  data?: Tutorial;
}

// 创建教程
export const createTutorial = async (
  params: CreateTutorialParams
): Promise<TutorialResult> => {
  try {
    const body = {
      title: params.title.trim(),
      desc: params.description.trim(),
      content: params.content.trim(),
      author: params.author.trim(),
      source_link: params.source_link,
      cover_img: params.cover_img,
      tags: params.tags ?? [],
      dapp_id: params.dapp_id,
    };

    const response = await apiRequest<TutorialResult>(
      '/tutorials',
      'POST',
      body
    );

    if (response.code === 200 && response.data) {
      return {
        success: true,
        message: '教程创建成功',
        data: response.data as unknown as Tutorial,
      };
    }

    return { success: false, message: '教程创建失败' };
  } catch (error: any) {
    console.error('创建教程异常:', error);
    return {
      success: false,
      message: error?.message ?? '网络错误，请稍后重试',
    };
  }
};

// 更新教程
export const updateTutorial = async (
  tutorialId: string,
  params: UpdateTutorialParams
): Promise<TutorialResult> => {
  try {
    const body = {
      title: params.title.trim(),
      desc: params.description.trim(),
      content: params.content.trim(),
      author: params.author.trim(),
      source_link: params.source_link,
      cover_img: params.cover_img,
      tags: params.tags ?? [],
      dapp_id: params.dapp_id,
    };

    const response = await apiRequest<TutorialResult>(
      `/tutorials/${tutorialId}`,
      'PUT',
      body
    );

    if (response.code === 200 && response.data) {
      return {
        success: true,
        message: response.message ?? '教程更新成功',
        data: response.data as unknown as Tutorial,
      };
    }

    return { success: false, message: response.message ?? '教程更新失败' };
  } catch (error: any) {
    return {
      success: false,
      message: error?.message ?? '网络错误，请稍后重试',
    };
  }
};

// 获取教程列表
export const getTutorials = async (
  params: GetTutorialsParams = {}
): Promise<TutorialListResult> => {
  try {
    const query = new URLSearchParams();

    if (params.keyword?.trim()) query.append('keyword', params.keyword.trim());
    if (params.tag?.trim()) query.append('tag', params.tag.trim());
    if (params.publish_status != null)
      query.append('publish_status', params.publish_status.toString());
    if (params.dapp_id != null)
      query.append('dapp_id', params.dapp_id.toString());
    if (params.user_id != null)
      query.append('user_id', params.user_id.toString());

    query.append('order', params.order ?? 'desc');
    query.append('page', (params.page ?? 1).toString());
    query.append('page_size', (params.page_size ?? 6).toString());

    const response = await apiRequest<TutorialListResult>(
      `/tutorials?${query.toString()}`,
      'GET'
    );

    if (response.code === 200 && response.data) {
      return {
        success: true,
        message: response.message ?? '获取教程列表成功',
        data: response.data as unknown as PaginatedTutorialData,
      };
    }

    return { success: false, message: response.message ?? '获取教程列表失败' };
  } catch (error: any) {
    console.error('获取教程列表异常:', error);
    return {
      success: false,
      message: error?.message ?? '网络错误，请稍后重试',
    };
  }
};

// 根据 ID 获取教程
export const getTutorialById = async (
  tutorialId: string
): Promise<TutorialResult> => {
  try {
    if (!tutorialId) {
      return { success: false, message: '教程ID不能为空' };
    }

    const response = await apiRequest<TutorialResult>(
      `/tutorials/${tutorialId}`,
      'GET'
    );

    if (response.code === 200 && response.data) {
      return {
        success: true,
        message: response.message ?? '获取教程成功',
        data: response.data as unknown as Tutorial,
      };
    }

    return { success: false, message: response.message ?? '获取教程失败' };
  } catch (error: any) {
    console.error('获取教程异常:', error);
    return {
      success: false,
      message: error?.message ?? '网络错误，请稍后重试',
    };
  }
};

// 删除教程
export const deleteTutorial = async (
  tutorialId: number
): Promise<TutorialResult> => {
  try {
    const response = await apiRequest<TutorialResult>(
      `/tutorials/${tutorialId}`,
      'DELETE'
    );

    if (response.code === 200) {
      return { success: true, message: response.message ?? '删除成功' };
    }

    return { success: false, message: response.message ?? '删除失败' };
  } catch (error: any) {
    console.error('删除教程异常:', error);
    return {
      success: false,
      message: error?.message ?? '网络错误，请稍后重试',
    };
  }
};

// 更新教程发布状态
export const updateTutorialPublishStatus = async (
  tutorialId: string,
  publishStatus: number
): Promise<TutorialResult> => {
  try {
    const body = {
      publish_status: publishStatus,
    };

    const response = await apiRequest<TutorialResult>(
      `/tutorials/${tutorialId}/status`,
      'PUT',
      body
    );

    if (response.code === 200 && response.data) {
      return {
        success: true,
        message: response.message ?? '教程状态更新成功',
        data: response.data as unknown as Tutorial,
      };
    }

    return { success: false, message: response.message ?? '教程状态更新失败' };
  } catch (error: any) {
    return {
      success: false,
      message: error?.message ?? '网络错误，请稍后重试',
    };
  }
};
